### PR TITLE
Minimal PHP version is 5.4.0

### PR DESCRIPTION
**Pull request by milo**
Maybe 5.4.6 (see bugs sanitization in tests) or even higher?
If agreed, I'll squash it all under the 1st commit.

What about to release 1.5.1 with all commits until Oct 1, 2015 and open the 2.0 version?

### CODE REVIEW

**Comment by Majkl578** [comment]
Hmm, is there a reason for keeping compatibility with unsupported PHP 5.4 & 5.5?
(I.e. PHPUnit 5 - released today - requires 5.6 & 7.0 only.)

**Comment by milo** [comment]
I have no idea, how is the 5.6 spreaded. Tester itself does not need 5.6 features necessarily.

Personaly, I'm ok with 5.6 for my projects.

**Comment by dg** [comment]
I would merge it as late as possible.

**Comment by dg** [comment]
Nette 2.3 will be actively maintained to february 2016. After that I will have no project developed under PHP 5.3, so I will not need Tester for 5.3. It does not mean that it is necessary to wait a month, it can be merged sooner.

**Comment by dg** [comment]
If you develop a library for PHP 5.4.0, the testing tool should not limit you and allow you to test it in PHP 5.4.0, so I think that minimal required version should be 5.*.0. And IMHO 5.5/5.6 has too few benefits over 5.4, so minimal required version 5.4.0 seems best to me.

File: src/CodeCoverage/Generators/template.phtml [hunk 1]
```diff
@@ -146,7 +146,7 @@
 	<?php endforeach ?>
 
 	<footer>
-		<p>Generated by <a href="https://tester.nette.org">Nette Tester</a> at <?php echo @date('Y/m/d H:i:s') // @ timezone may not be set ?></p>
+		<p>Generated by <a href="http://tester.nette.org">Nette Tester</a> at <?= @date('Y/m/d H:i:s') // @ timezone may not be set ?></p>
```
Comment (dg) [inline src/CodeCoverage/Generators/template.phtml]: Here should stay https

File: src/Framework/Assert.php [hunk 2]
```diff
@@ -283,7 +283,7 @@ public static function type($type, $value, $description = NULL)
 	 * @param  int code
 	 * @return \Exception
 	 */
-	public static function exception($function, $class, $message = NULL, $code = NULL)
+	public static function exception(callable $function, $class, $message = NULL, $code = NULL)
```
Comment (dg) [inline src/Framework/Assert.php]: … and method Assert::throws()

File: tests/Runner/Runner.edge.phpt [hunk 3]
```diff
@@ -35,7 +31,7 @@ $cli = PHP_SAPI === 'cli';
 $bug62725 = $cli && PHP_VERSION_ID >= 50400 && PHP_VERSION_ID <= 50406;
 Assert::same($bug62725 ? [Runner::PASSED, NULL] : [Runner::FAILED, 'Exited with error code 231 (expected 0)'], $logger->results['shutdown.exitCode.a.phptx']);
 
-$bug65275 = !defined('HHVM_VERSION') && $cli && PHP_VERSION_ID >= 50300; // bug still exists
+$bug65275 = !defined('HHVM_VERSION') && $cli;
 Assert::same($bug65275 ? [Runner::FAILED, 'Exited with error code 231 (expected 0)'] : [Runner::PASSED, NULL], $logger->results['shutdown.exitCode.b.phptx']);
```
Comment (dg) [inline tests/Runner/Runner.edge.phpt]: Are you sure here?

File: src/Framework/Dumper.php [hunk 4]
```diff
@@ -307,7 +307,7 @@ public static function dumpException($e)
 			. (isset($stored) ? 'diff ' . Helpers::escapeArg($stored[0]) . ' ' . Helpers::escapeArg($stored[1]) . "\n\n" : '');
 
 		foreach ($trace as $item) {
-			$item += array('file' => NULL, 'class' => NULL, 'type' => NULL, 'function' => NULL);
+			$item += ['file' => NULL, 'class' => NULL, 'type' => NULL, 'function' => NULL];
 			if ($e instanceof AssertException && $item['file'] === __DIR__ . DIRECTORY_SEPARATOR . 'Assert.php') {
 				continue;
 			}
```
Comment (dg) [inline src/Framework/Dumper.php]: The line bellow can be simplified to `&& strpos($tmp = file($item['file'])[$item['line'] - 1], "::$item[function](") ? $tmp : NULL;`

**Comment by dg** [comment]
> If agreed, I'll squash it all under the 1st commit.

IMHO more than one commit is better, I used one commit on other repositories and it was not good.

File: tests/Runner/Runner.edge.phpt (hunk 3 above)
Comment (milo) [inline tests/Runner/Runner.edge.phpt]: I cannot figure how the removed part could ever be FALSE with PHP 5.3+.

File: src/Framework/Dumper.php (hunk 4 above)
Comment (milo) [inline src/Framework/Dumper.php]: Thanks, that is I couldn't recall for :)

File: tests/Runner/Runner.edge.phpt (hunk 3 above)
Comment (dg) [inline tests/Runner/Runner.edge.phpt]: You right

File: src/Framework/Dumper.php (hunk 4 above)
Comment (dg) [inline src/Framework/Dumper.php]: This is the most cryptic part of Tester :-)

**Comment by milo** [comment]
>> If agreed, I'll squash it all under the 1st commit.

> IMHO more than one commit is better, I used one commit on other repositories and it was not good.

OK. Plan is rebase it, update it, maybe squash some commits together.

File: src/Framework/Environment.php [hunk 5]
```diff
@@ -70,10 +70,9 @@ public static function setupColors()
 				&& ((function_exists('posix_isatty') && posix_isatty(STDOUT))
 					|| getenv('ConEmuANSI') === 'ON' || getenv('ANSICON') !== FALSE) || getenv('term') === 'xterm-256color');
 
-		$colors = & self::$useColors;
-		ob_start(function ($s) use (& $colors) {
-			return $colors ? $s : Dumper::removeColors($s);
-		}, PHP_VERSION_ID < 50400 ? 2 : 1, FALSE);
+		ob_start(function ($s) {
```
Comment (dg) [inline src/Framework/Environment.php]: This should probably stay https://github.com/nette/tester/commit/d76cd9776f757763ce1479602c1c3232c95c0a81

File: src/CodeCoverage/Generators/AbstractGenerator.php [hunk 6]
```diff
@@ -96,31 +96,13 @@ protected function getSourceIterator()
 			? new \RecursiveIteratorIterator(new \RecursiveDirectoryIterator($this->source))
 			: new \ArrayIterator([new \SplFileInfo($this->source)]);
 
-		return new AcceptIterator($iterator, $this->acceptFiles);
+		return new \CallbackFilterIterator($iterator, function (\SplFileInfo $file) {
+			return substr($file->getBasename(), 0, 1) !== '.'  // . or .. or .gitignore
```
Comment (JanTvrdik) [inline src/CodeCoverage/Generators/AbstractGenerator.php]: `$file->getBasename()[0]` should work fine assuming `getBasename()` cannot be empty

**Comment by milo** [comment]
Updated. Imho, ready to merge. Some other objectives or updates?

**Comment by milo** [comment]
GitHub closed the PR when I accidentally pushed `master` into `pull-php54` branch. Pushing commits back didn't help and PR now looks like 0 commits.

Merged:
- 077c3a940c26e251517a9f94d20b6d5a5985c39f
- db9688f40b666276579903e19ac852a2373354ac
- d7aad9ea7bcb9f06605d4c0b4d3a7c05b336001d
- 33fb94a711c4bba0a77287ca9904b04585bf48c4
- f836dbbe2d304d9fb687ab5a9148589cc5c41791
- c1bdc6fc106efb820fa901e19598d28f6a45efd9
- 8bd9db7f1c4a2e3acabeabcdf94c361009134fe8

**Pull request closed without merging**